File: lintcode/binary_search/First Bad Version.py
#class SVNRepo:
#    @classmethod
#    def isBadVersion(cls, id)
#        # Run unit tests to check whether verison `id` is a bad version
#        # return true if unit tests passed else false.
# You can use SVNRepo.isBadVersion(10) to check whether version 10 is a 
# bad version.
class Solution:
    """
    @param n: An integer
    @return: An integer which is the first bad version.
    """
    def findFirstBadVersion(self, n):
        # write your code here
        b, e = 1, n
        while b + 1 < e:
            mid = b + (e-b) // 2
            if SVNRepo.isBadVersion(mid):
                e = mid
            else:
                b = mid
        if SVNRepo.isBadVersion(b):
            return b
        return e

File: leetcode/binarySearch/firstBadVersion.go
package binarySearch

/**
 * Forward declaration of isBadVersion API.
 * @param   version   your guess about first bad version
 * @return 	 	      true if current version is bad
 *			          false if current version is good
 * func isBadVersion(version int) bool;
 */

func firstBadVersion(n int) int {
	b, e := 1, n
	for; b+1 < e;{
		m := b + (e-b)/2
		if isBadVersion(m){
			e = m
		}else{
			b = m
		}
	}
	if isBadVersion(b){
		return b
	}
	return e
}

